client.go

```diff
@@ -249,9 +311,7 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 
 		// Always rewind the request body when non-nil.
 		if req.body != nil {
-			if _, err := req.body.Seek(0, 0); err != nil {
-				return nil, fmt.Errorf("failed to seek body: %v", err)
-			}
+			req.Request.Body = ioutil.NopCloser(req.body())
```

Comment:
I think we need to put the `NopCloser` wrapper specifically in the `ReadSeeker` case only. We want it to close any other kind of reader.

Reply:
I had it like that but wasn't really sure. The reason is that if someone passes in a func that returns an io.Reader, that io.Reader may also implement a Close method so the check in net/http for whether it satisfies io.Closer will pass. IOW without knowing the full spectrum of things that might be passed in, this seemed safer. But we could leave that up to the user to sort out.

Reply:
Yeah, I think with the supported types we have here, the only time we will want to prevent the `Close()` will be for the `ReadSeeker`, since the types from the `bytes` package don't implement a `Close()`, and the function type definitely needs to have its readers closed if they support it.

Just for example's sake, if I pass a function that returns an `os.File`, that will never be closed and leak an FD if we mask the `Close()` method.

Reply:
Yep, sounds good, will fix.